### PR TITLE
Add rows_before_aggregation_at_least statistic

### PR DESCRIPTION
**Pull request by morning-color**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Add the `rows_before_aggregation_at_least` statistic to the query response when `rows_before_aggregation` is enabled. This statistic represents the number of rows read before aggregation.
In the context of a distributed query, when using the `group by` or `max` aggregation function without a `limit`, `rows_before_aggregation_at_least` can reflect the number of rows hit by the query.

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll3** [comment]
*This is an automated comment for commit 5f42e1518277e54c34fce274b5cf46cd5069cb2a with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/66084/5f42e1518277e54c34fce274b5cf46cd5069cb2a/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

File: src/QueryPipeline/QueryPipeline.cpp [hunk 1]
```diff
@@ -273,7 +274,28 @@ static void initRowsBeforeLimit(IOutputFormat * output_format)
...
+static void initRowsBeforeAggregation(std::shared_ptr<Processors> processors, IOutputFormat * output_format)
+{
+    bool has_aggregation = false;
 
+    if (!processors->empty())
+    {
+        RowsBeforeAggregationCounterPtr rows_before_aggregation_at_least = std::make_shared<RowsBeforeStepCounter>();
+        for (auto processor : *processors)
+        {
+            if (auto transform = std::dynamic_pointer_cast<AggregatingTransform>(processor))
```
Comment (CurtizJ) [inline src/QueryPipeline/QueryPipeline.cpp:286]: Also take into account `AggregatingInOrderTransform`.

File: src/Processors/Sources/RemoteSource.cpp [hunk 2]
```diff
@@ -35,16 +35,25 @@ RemoteSource::RemoteSource(RemoteQueryExecutorPtr executor, bool add_aggregation
...
+                else
+                    manually_add_rows_before_limit_counter = true; /// Remote subquery doesn't contain a limit
+            }
+
+            if (rows_before_aggregation)
+            {
+                if (info.hasAppliedAggregation())
+                    rows_before_aggregation->add(info.getRowsBeforeAggregation());
+                else
+                    manually_add_rows_before_aggregation_counter = true; /// Remote subquery doesn't contain a group by
```
Comment (CurtizJ) [inline src/Processors/Sources/RemoteSource.cpp:54]: Looks like it's the reason why your test fails with remote execution. You count the same rows twice: here and in `AggregatingTransform`. Probably you don't need to add rows manually in that case.

**Review by CurtizJ — COMMENTED** [review]
Actually you don't need the suffix `at_least` because number of rows before aggregation is always exact. The only case when it's not true is when you have mixed version cluster and some of the nodes don't report this statistic.

File: src/Processors/Formats/IOutputFormat.h [hunk 3]
```diff
@@ -36,11 +36,17 @@ class IOutputFormat : public IProcessor
...
 
     /// Value for rows_before_limit_at_least field.
-    virtual void setRowsBeforeLimit(size_t /*rows_before_limit*/) {}
+    virtual void setRowsBeforeLimit(size_t /*rows_before_limit*/) { }
 
     /// Counter to calculate rows_before_limit_at_least in processors pipeline.
     void setRowsBeforeLimitCounter(RowsBeforeLimitCounterPtr counter) override { rows_before_limit_counter.swap(counter); }
 
+    /// Value for rows_before_aggregation_at_least field.
+    virtual void setRowsBeforeAggregation(size_t /*rows_before_limit*/) { }
```
[the remark this answers is not in the document]
Comment (morning-color) [reply src/Processors/Formats/IOutputFormat.h:45]: Sure, I will rename it.

File: src/QueryPipeline/ProfileInfo.cpp [hunk 4]
```diff
@@ -16,6 +16,8 @@ void ProfileInfo::read(ReadBuffer & in)
     readBinary(applied_limit, in);
     readVarUInt(rows_before_limit, in);
     readBinary(calculated_rows_before_limit, in);
+    readBinary(applied_aggregation, in);
+    readVarUInt(rows_before_aggregation, in);
```
[the remark this answers is not in the document]
Comment (morning-color) [reply src/QueryPipeline/ProfileInfo.cpp:20]: Thank you very much for your reminder. I will add logic to ensure compatibility with lower old client versions.

File: src/Processors/IProcessor.h [hunk 5]
```diff
@@ -21,8 +21,10 @@ class IQueryPlanStep;
 struct StorageLimits;
 using StorageLimitsList = std::list<StorageLimits>;
 
-class RowsBeforeLimitCounter;
-using RowsBeforeLimitCounterPtr = std::shared_ptr<RowsBeforeLimitCounter>;
+class RowsBeforeStepCounter;
+using RowsBeforeLimitCounterPtr = std::shared_ptr<RowsBeforeStepCounter>;
+
+using RowsBeforeAggregationCounterPtr = std::shared_ptr<RowsBeforeStepCounter>;
```
[the remark this answers is not in the document]
Comment (morning-color) [reply src/Processors/IProcessor.h:28]: Sure, I will rename it.

File: src/QueryPipeline/QueryPipeline.cpp (hunk 1 above)
Comment (morning-color) [reply src/QueryPipeline/QueryPipeline.cpp:286]: Thank you very much for your reminder. I will take it into account.

File: src/Processors/Sources/RemoteSource.cpp (hunk 2 above)
Comment (morning-color) [reply src/Processors/Sources/RemoteSource.cpp:54]: Oh, I see. I will fix it as soon as possible and submit the commit.

File: src/QueryPipeline/QueryPipeline.cpp [hunk 6]
```diff
@@ -273,7 +276,28 @@ static void initRowsBeforeLimit(IOutputFormat * output_format)
...
     }
 }
+static void initRowsBeforeAggregation(std::shared_ptr<Processors> processors, IOutputFormat * output_format)
+{
+    bool has_aggregation = false;
 
+    if (!processors->empty())
+    {
+        RowsBeforeStepCounterPtr rows_before_aggregation = std::make_shared<RowsBeforeStepCounter>();
+        for (auto processor : *processors)
```
Comment (CurtizJ) [inline src/QueryPipeline/QueryPipeline.cpp:286]:
```suggestion
        for (const auto & processor : *processors)
```

File: src/QueryPipeline/QueryPipeline.cpp [hunk 7]
```diff
@@ -521,6 +545,14 @@ void QueryPipeline::complete(std::shared_ptr<IOutputFormat> format)
     extremes = nullptr;
 
     initRowsBeforeLimit(format.get());
+    for (const auto context : resources.interpreter_context)
```
Comment (CurtizJ) [inline src/QueryPipeline/QueryPipeline.cpp:548]:
```suggestion
    for (const auto & context : resources.interpreter_context)
```

**Review by CurtizJ — COMMENTED** [review]
clang-tidy has failed

**Review by CurtizJ — APPROVED** [review]
LGTM

**Pull request merged**